pipeline_dp/pld_budget_accounting.py

```diff
@@ -0,0 +1,114 @@
+    total_epsilon: float = None
+    total_delta: float = None
+    _requested_laplace_budgets: list = field(default_factory=list)
+    _requested_gaussian_budgets: list = field(default_factory=list)
+
+    def __post_init__(self):
+        if self.total_epsilon is None or self.total_delta is None:
+            raise ValueError(f"Must pass total_epsilon and total_delta")
+        if self.total_epsilon is not None and self.total_epsilon <= 0:
+            raise ValueError(
+                f"Epsilon must be positive, not {self.total_epsilon}.")
+        if self.total_delta is not None and self.total_delta < 0:
+            raise ValueError(
+                f"Delta must be non-negative, not {self.total_delta}.")
+        if self.noise > self.max_noise:
+            raise ValueError(
+                f"Noise = {self.noise} is larger than the maximum = {self.max_noise}."
+            )
+
+    def request_budget(self, noise_kind: NoiseKind) -> PLDBudget:
+        pldBudget = PLDBudget()
+        if noise_kind == NoiseKind.LAPLACE:
+            self._requested_laplace_budgets.append(
+                RequestedPLDBudget(pldBudget))
+        elif noise_kind == NoiseKind.GAUSSIAN:
+            self._requested_gaussian_budgets.append(
+                RequestedPLDBudget(pldBudget))
+        else:
+            raise NotImplementedError(
+                f"Noise type {noise_kind} is unsupported.")
+        return pldBudget
+
+    def compute_budgets(self):
+        optimal_noise = self.optimize_noise_to_delta()
+        self.noise = optimal_noise  # i set the noise for the entire budget accountant
+
+    def optimize_noise_to_delta(self) -> float:
+        """Adjust noise to get close to given epsilon; increasing noise decreases epsilon."""
+        threshold = 0.001
+        optimal_noise = self.noise
+        pld_laplace_gaussian = self.compose_distributions(optimal_noise)
+        pld_epsilon = pld_laplace_gaussian.get_epsilon_for_delta(
+            self.total_delta)
+        (min, max) = (0, self.max_noise)
+
+        while (min <= max):
+            if abs(pld_epsilon - self.total_epsilon) <= threshold:
+                return optimal_noise
+            elif pld_epsilon < self.total_epsilon:
+                max = optimal_noise
+            elif pld_epsilon > self.total_epsilon:
+                min = optimal_noise
+
+            optimal_noise = ((max - min) / 2) + min
+            pld_laplace_gaussian = self.compose_distributions(optimal_noise)
+            pld_epsilon = pld_laplace_gaussian.get_epsilon_for_delta(
+                self.total_delta)
+
+    def compose_distributions(self, noise: float):
+        pld_laplace = self.compose_laplace(noise)
+        pld_gaussian = self.compose_gaussian(noise)
+        if pld_gaussian is None:
+            return pld_laplace
+        if pld_gaussian is None:
+            return pld_gaussian
+        return pld_laplace.compose(pld_gaussian)
+
+    def compose_laplace(self, noise: float):
```

Comment:
the difference is in compose_laplace and compose_gaussian is only in creating. And it's not necessary to compose laplace and gaussians separately, so I'd suggest:
1.Have only 1 var _requested_budgets 
2.To do composition in one loop 

Clarification: In such way we would remove code duplication and simplify code. Moreover in future we will have more mechanisms and so it's better to keep all of them in one variable _requested_budgets.

Reply:
👍  Fixed to loop around a single collection, thank you!